### PR TITLE
Fix MaxSat2TypeSolver

### PR DESCRIPTION
**Pull request by d367wang**
Fixes https://github.com/opprop/checker-framework-inference/issues/328

### CODE REVIEW

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 1]
```diff
@@ -108,30 +118,38 @@ public InferenceResult solve() {
...
 
                 for (Integer var : solution) {
-                    boolean isTop = var < 0;
-                    if (isTop) {
-                        var = -var;
-                    }
-
                     Integer potential = existentialToPotentialIds.get(var);
                     if (potential != null) {
                         // idToExistence.put(potential, !isTop);
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:123]: remove

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 2]
```diff
@@ -108,30 +118,38 @@ public InferenceResult solve() {
...
 
                 }
             } else {
                 System.out.println("Not solvable!");
             }
 
+        } catch (ContradictionException ce) {
+            System.out.println("Find contradiction when adding " + lastClause + ". Not solvable!");
+            // pass empty set as the unsat explanation
+            // TODO: explain UNSAT
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:145]: try reusing MaxSATUnsatisfiableConstraintExplainer

File: src/checkers/inference/model/serialization/CnfVecIntSerializer.java [hunk 3]
```diff
@@ -288,12 +289,37 @@ public boolean emptyClause(VecInt ... clauses) {
...
     }
 
+    /**
+     * Convert all the given mandatory constraints into hard clauses. A BugInCF exception is
+     * raised if the given constraints contains any {@link PreferenceConstraint}.
+     *
+     * For conversion of constraints containing {@link PreferenceConstraint}, use
+     * {@link CnfVecIntSerializer#convertAll(Iterable, List, List)}
+     *
+     * @param constraints
```
Comment (wmdietl) [inline src/checkers/inference/model/serialization/CnfVecIntSerializer.java:299]:
```suggestion
     * @param constraints the constraints to convert
```

File: src/checkers/inference/model/serialization/CnfVecIntSerializer.java [hunk 4]
```diff
@@ -288,12 +289,37 @@ public boolean emptyClause(VecInt ... clauses) {
...
```
Comment (wmdietl) [inline src/checkers/inference/model/serialization/CnfVecIntSerializer.java:313]:
```suggestion
     * @param constraints the constraints to convert
```

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 5]
```diff
@@ -108,30 +118,45 @@ public InferenceResult solve() {
...
             }
 
+        } catch (ContradictionException ce) {
+            // This case indicates that constraints are not solvable, too. This is normal so continue
+            // the execution
+            System.out.println("Not solvable! Contradiction exception " +
+                    "when adding clause: " + lastClause + ".");
+
+            // pass empty set as the unsat explanation
+            // TODO: explain UNSAT
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:153]:
```suggestion
            // TODO: explain UNSAT possibly by reusing MaxSatSolver.MaxSATUnsatisfiableConstraintExplainer
```

File: src/checkers/inference/model/serialization/CnfVecIntSerializer.java [hunk 6]
```diff
@@ -288,12 +289,37 @@ public boolean emptyClause(VecInt ... clauses) {
...
+     * @param results the output clauses for the given constraints
+     * @return same as {@code results}
+     */
     public List<VecInt> convertAll(Iterable<Constraint> constraints, List<VecInt> results) {
         for (Constraint constraint : constraints) {
+            if (constraint instanceof PreferenceConstraint) {
+                throw new BugInCF("CnfVecIntSerializer: adding PreferenceConstraint ( " + constraint +
+                        " ) to the hard clauses");
```
Comment (wmdietl) [inline src/checkers/inference/model/serialization/CnfVecIntSerializer.java:321]:
```suggestion
                        " ) to hard clauses is forbidden");
```

File: src/checkers/inference/model/serialization/CnfVecIntSerializer.java [hunk 7]
```diff
@@ -304,6 +330,28 @@ public boolean emptyClause(VecInt ... clauses) {
         return results;
     }
 
+    /**
+     * Convert all the given mandatory constraints to hard clauses, and preference constraints
+     * to soft clauses.
+     *
+     * @param constraints
```
Comment (wmdietl) [inline src/checkers/inference/model/serialization/CnfVecIntSerializer.java:337]:
```suggestion
     * @param constraints the constraints to convert
```

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 8]
```diff
@@ -108,30 +118,45 @@ public InferenceResult solve() {
...
+                        boolean isTop = var < 0;
+                        if (isTop) {
+                            var = -var;
+                        }
+                        solutions.put(var, isTop ? top : bottom);
                     }
 
                 }
             } else {
+                // TODO: explain UNSAT
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:142]:
```suggestion
                // TODO: explain UNSAT possibly by reusing MaxSatSolver.MaxSATUnsatisfiableConstraintExplainer
```

File: src/checkers/inference/model/serialization/CnfVecIntSerializer.java [hunk 9]
```diff
@@ -288,12 +289,37 @@ public boolean emptyClause(VecInt ... clauses) {
                 "MaxSAT backend. Use MaxSATSolver instead!");
     }
 
+    /**
+     * Convert all the given mandatory constraints into hard clauses. A BugInCF exception is
+     * raised if the given constraints contains any {@link PreferenceConstraint}.
```
Comment (d367wang) [inline src/checkers/inference/model/serialization/CnfVecIntSerializer.java:294]:
```suggestion
     * raised if the given constraints contain any {@link PreferenceConstraint}.
```

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 10]
```diff
@@ -108,30 +118,45 @@ public InferenceResult solve() {
...
-                    }
-
-                    Integer potential = existentialToPotentialIds.get(var);
+                    Integer potential = existentialToPotentialIds.get(Math.abs(var));
                     if (potential != null) {
-                        // idToExistence.put(potential, !isTop);
-                        // TODO: which AnnotationMirror should be used?
-                        solutions.put(potential, bottom);
+                        // Assume the 'solution' output by the solver is already sorted in the ascending order
+                        // of their absolute values. So the existential variables come after the slot variables,
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:124]:
```suggestion
                        // of their absolute values. So the existential variables come after the potential variables,
```

File: src/checkers/inference/solver/MaxSat2TypeSolver.java [hunk 11]
```diff
@@ -108,30 +118,45 @@ public InferenceResult solve() {
...
+                        // which means the potential slot corresponding to the current existential variable is
+                        // already inserted into 'solutions'
+                        assert solutions.containsKey(potential);
+                        if (var < 0) {
+                            // potential variable does not exist, remove it from the solution
```
Comment (d367wang) [inline src/checkers/inference/solver/MaxSat2TypeSolver.java:129]:
```suggestion
                            // The existential variable is false, so the potential variable should not be inserted. 
                            // Remove it from the solution.
```

**Review by wmdietl — APPROVED** [review]

**Pull request merged**